### DISCUSSION TITLE
End user firmware update

### DISCUSSION THREAD

**Discussion by ht93** [stm32duino/Arduino_Core_STM32#1717] [Q&A]
Hi,

I am working on a project with stm32f407 and stm32duino, and I am using st link to uploading sketch. I have questions about end user firmware update:
1. I want end user to able do the firmware update without the st link, arduino IDE and sketch. I think USB (maybe via USART) or SD card would be most convenient for them. How can I do firmware update with current 2.2.0 stm32duino core through USB or SD card?
2. If I use USB, I probably need to write a program in PC to upload the firmware (i.e. without the st link, arduino IDE and sketch, how can I do it?
3. If I use SD card for firmware update, can I store data like user data (like settings that they changed) also in the SD card at same time? How can I do it?

Thanks a lot!

**Comment by fpistm** [comment]
STM32F407 probably support DFU upload. User have simply to enter properly in bootloader mode with USB connected.
See §28.1.1 Bootloader configuration:
https://www.st.com/resource/en/application_note/cd00167594-stm32-microcontroller-system-memory-boot-mode-stmicroelectronics.pdf

So pattern 1 should be used:
Pattern 1 Boot0(pin) = 1 and Boot1(pin) = 0

Then select STM32CubeProgrammer(DFU) upload method.

**Reply by ht93** [reply]
Thanks for reply!

Looks like there are two parts:
1. Enter bootloader mode by pattern 1 (Boot0(pin) = 1 and Boot1(pin) = 0) and connect with USB. For this part, the end user must need to change the pin state of boot0 and boot1 right? Is it possible to update firmware with only connecting usb cable and let the program in PC force the firmware update (without the need to push any button)?
2. Select "STM32CubeProgrammer(DFU) upload method" means use arduino IDE or STM32CubeProgrammer to upload firmware? For end user, is it possible to write some standalone program for upload firmware without the arduino IDE and STM32CubeProgrammer

**Reply by fpistm** [reply]
> 1. Enter bootloader mode by pattern 1 (Boot0(pin) = 1 and Boot1(pin) = 0) and connect with USB. For this part, the end user must need to change the pin state of boot0 and boot1 right? Is it possible to update firmware with only connecting usb cable and let the program in PC force the firmware update (without the need to push any button)?

Currently no. If your boards has jumper or push button for those pins it is the easiest way. I've made a first [POC](https://github.com/stm32duino/Arduino_Core_STM32/pull/710) to jump directly into the BL mode but had no time to end it.



> 2\. Select "STM32CubeProgrammer(DFU) upload method" means use arduino IDE or STM32CubeProgrammer to upload firmware? For end user, is it possible to write some standalone program for upload firmware without the arduino IDE and STM32CubeProgrammer

Yes. Arduino uses STM32Cube programmer installed on the host as it cannot be delivered as an Arduino package. Other alternative is to use DFU utils binary but currently not available for all boards.

**Reply by ht93** [reply]
I see. Until you finish [POC](https://github.com/stm32duino/Arduino_Core_STM32/pull/710), it is required to have button pressed or jumper for uploading firmware through USB right?

If I use SD card, instead of USB, for firmware update (user just copy the new firmware bin file to SD card specify location), is it possible to load the firmware from SD card? As user probably can choose to update the firmware with device normal IO (some button and rotary encoder for normal control function).

**Reply by fpistm** [reply]
It should be possible but in that case you will have to write the proper sequence to flash it properly. I'm not aware of example on this use case.

**Reply by ht93** [reply]
I do find some example for stm32 sd card firmware update but not with stm32duino core, like https://www.youtube.com/watch?v=v_JxDAiIDyc and https://github.com/akospasztor/stm32-bootloader. But I am not sure how to adapt them to stm32duino core.

**Reply by fpistm** [reply]
Unfortunately, I will not be able to help on this. You could try to get help on the forum.

**Reply by ht93** [reply]
@fpistm Hi, if I connect boot1 pin to ground, and boot0 pin to the VCC of the usb port. It will enter the DFU mode if I plug in usb cable (other side plugged to PC) right?

**Reply by fpistm** [reply]
Yes it should
